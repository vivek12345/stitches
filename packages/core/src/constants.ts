export const tokenTypes = [
  'colors',
  'space',
  'fontSizes',
  'fonts',
  'fontWeights',
  'lineHeights',
  'letterSpacings',
  'sizes',
  'borderWidths',
  'borderStyles',
  'radii',
  'shadows',
  'zIndices',
  'transitions',
] as const;
// Note: when running Jest tests, make sure that the test file is running in node env
// if this constant was giving incorrect results.
export const isServer = typeof window === 'undefined';
