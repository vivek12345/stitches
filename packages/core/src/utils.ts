import {
  animation,
  background,
  border,
  borderBottom,
  borderColor,
  borderLeft,
  borderRadius,
  borderRight,
  borderStyle,
  borderTop,
  borderWidth,
  boxShadow,
  font,
  gap,
  margin,
  padding,
  transition,
  textDecoration,
  flex,
} from './shorthand-parser';
import { IBreakpoints, ICssPropToToken, ISheet } from './types';

export const MAIN_BREAKPOINT_ID = 'initial';
export type TMainBreakPoint = typeof MAIN_BREAKPOINT_ID;

export const cssPropToToken: ICssPropToToken<any> = {
  gap: 'space',
  gridGap: 'space',
  columnGap: 'space',
  gridColumnGap: 'space',
  rowGap: 'space',
  gridRowGap: 'space',
  inset: 'space',
  insetBlock: 'space',
  insetBlockEnd: 'space',
  insetBlockStart: 'space',
  insetInline: 'space',
  insetInlineEnd: 'space',
  insetInlineStart: 'space',
  margin: 'space',
  marginTop: 'space',
  marginRight: 'space',
  marginBottom: 'space',
  marginLeft: 'space',
  marginBlock: 'space',
  marginBlockEnd: 'space',
  marginBlockStart: 'space',
  marginInline: 'space',
  marginInlineEnd: 'space',
  marginInlineStart: 'space',
  padding: 'space',
  paddingTop: 'space',
  paddingRight: 'space',
  paddingBottom: 'space',
  paddingLeft: 'space',
  paddingBlock: 'space',
  paddingBlockEnd: 'space',
  paddingBlockStart: 'space',
  paddingInline: 'space',
  paddingInlineEnd: 'space',
  paddingInlineStart: 'space',
  top: 'space',
  right: 'space',
  bottom: 'space',
  left: 'space',
  fontSize: 'fontSizes',
  backgroundColor: 'colors',
  border: ['', 'borderStyles', 'colors'],
  borderColor: 'colors',
  borderTopColor: 'colors',
  borderRightColor: 'colors',
  borderBottomColor: 'colors',
  borderLeftColor: 'colors',
  caretColor: 'colors',
  color: 'colors',
  columnRuleColor: 'colors',
  outlineColor: 'colors',
  fill: 'colors',
  stroke: 'colors',
  fontFamily: 'fonts',
  fontWeight: 'fontWeights',
  lineHeight: 'lineHeights',
  letterSpacing: 'letterSpacings',
  blockSize: 'sizes',
  minBlockSize: 'sizes',
  maxBlockSize: 'sizes',
  inlineSize: 'sizes',
  minInlineSize: 'sizes',
  maxInlineSize: 'sizes',
  width: 'sizes',
  minWidth: 'sizes',
  maxWidth: 'sizes',
  height: 'sizes',
  minHeight: 'sizes',
  maxHeight: 'sizes',
  flexBasis: 'sizes',
  borderWidth: 'borderWidths',
  borderTopWidth: 'borderWidths',
  borderLeftWidth: 'borderWidths',
  borderRightWidth: 'borderWidths',
  borderBottomWidth: 'borderWidths',
  borderStyle: 'borderStyles',
  borderTopStyle: 'borderStyles',
  borderLeftStyle: 'borderStyles',
  borderRightStyle: 'borderStyles',
  borderBottomStyle: 'borderStyles',
  borderRadius: 'radii',
  borderTopLeftRadius: 'radii',
  borderTopRightRadius: 'radii',
  borderBottomRightRadius: 'radii',
  borderBottomLeftRadius: 'radii',
  boxShadow: 'shadows',
  textShadow: 'shadows',
  zIndex: 'zIndices',
  transition: 'transitions',
};

export const tokenTypes = [
  'sizes',
  'colors',
  'space',
  'fontSizes',
  'lineHeights',
  'fontWeights',
  'fonts',
  'borderWidths',
  'radii',
] as const;

const enhanceSheet = (sheet: ISheet): ISheet => {
  return {
    content: sheet.content,
    cssRules: sheet.cssRules,
    insertRule: (rule, index) => {
      try {
        sheet.insertRule(rule, index);
        return index;
      } catch {
        return -1;
      }
    },
  };
};

export const createSheets = (env: any, screens: IBreakpoints = {}) => {
  const tags: HTMLStyleElement[] = [];
  if (env && env.document) {
    const head = env.document.querySelector('head');

    if (!head) {
      throw new Error('There is no HEAD element on this document');
    }

    const styles = Array.from<HTMLStyleElement>(head.querySelectorAll('style'));
    const existingStyles = styles.filter((style) =>
      Boolean(style.textContent && style.textContent.startsWith('/* STITCHES'))
    );

    return {
      tags,
      sheets: ['__variables__', '__keyframes__', MAIN_BREAKPOINT_ID]
        .concat(Object.keys(screens))
        .reduce<{ [key: string]: ISheet }>((aggr, key, index) => {
          let style = existingStyles[index];
          if (!style) {
            style = env.document.createElement('style');
            head.appendChild(style);
          }
          tags.push(style);
          aggr[key] = enhanceSheet(style.sheet as any);

          return aggr;
        }, {}),
    };
  }

  return {
    tags,
    sheets: ['__variables__', '__keyframes__', MAIN_BREAKPOINT_ID]
      .concat(Object.keys(screens))
      .reduce<{ [key: string]: ISheet }>((aggr, key) => {
        aggr[key] = enhanceSheet({
          content: '',
          cssRules: [],
          insertRule(content, index = 0) {
            this.cssRules.splice(index, 0, content);
          },
        });

        return aggr;
      }, {}),
  };
};

export const specificityProps: {
  [key: string]: any;
} = {
  border,
  boxShadow,
  overflow: (tokens: any, value: any) => ({
    overflowX: value,
    overflowY: value,
  }),
  margin,
  padding,
  borderRadius,
  borderColor,
  borderStyle,
  borderWidth,
  background,
  animation,
  transition,
  font,
  borderBottom,
  borderLeft,
  borderTop,
  borderRight,
  textDecoration,
  gap,
  flex,
};

export const getVendorPrefixAndProps = (env: any) => {
  const styles = env.getComputedStyle(env.document.documentElement);
  const vendorProps = Array.from(styles).filter((prop) => (prop as string)[0] === '-');
  // @ts-ignore
  const vendorPrefix = (vendorProps.join('').match(/-(moz|webkit|ms)-/) || (styles.OLink === '' && ['', 'o']))[1];

  return { vendorPrefix: `-${vendorPrefix}-`, vendorProps };
};

export const hashString = (str: string) => {
  let hash = 5381;
  let i = str.length;

  while (i) {
    hash = (hash * 33) ^ str.charCodeAt(--i);
  }

  /* JavaScript does bitwise operations (like XOR, above) on 32-bit signed
   * integers. Since we want the results to be always positive, convert the
   * signed int to an unsigned by doing an unsigned bitshift. */
  return generateAlphabeticName(hash >>> 0);
};

/**
 * Converts a hash number to alphabetic representation:
 * Copied from:
 * https://github.com/styled-components/styled-components/blob/master/packages/styled-components/src/utils/generateAlphabeticName.js
 */

const AD_REPLACER_R = /(a)(d)/gi;

/* This is the "capacity" of our alphabet i.e. 2x26 for all letters plus their capitalised
 * counterparts */
const charsLength = 52;

/* start at 75 for 'a' until 'z' (25) and then start at 65 for capitalised letters */
const getAlphabeticChar = (code: number): string => String.fromCharCode(code + (code > 25 ? 39 : 97));

/* input a number, usually a hash and convert it to base-52 */
function generateAlphabeticName(code: number): string {
  let name = '';
  let x;

  /* get a char and divide by alphabet-length */
  for (x = Math.abs(code); x > charsLength; x = (x / charsLength) | 0) {
    name = getAlphabeticChar(x % charsLength) + name;
  }

  return (getAlphabeticChar(x % charsLength) + name).replace(AD_REPLACER_R, '$1-$2');
}
