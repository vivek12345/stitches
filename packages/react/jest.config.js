const base = require('../../jest.config.base.js');
const packageJson = require('./package.json');

module.exports = {
  ...base,
  roots: ['<rootDir>/src', '<rootDir>/tests'],
  displayName: packageJson.name,
};
