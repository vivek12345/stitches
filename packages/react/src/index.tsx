import {
  MAIN_BREAKPOINT_ID,
  TConfig,
  TCss,
  TCssProperties,
  TMainBreakPoint,
  createCss,
  hashString,
} from '@stitches/core';
export { _ATOM } from '@stitches/core';
import * as React from 'react';

let hasWarnedInlineStyle = false;

export type TCssProp<T extends TConfig> = TCssProperties<T> | (string & {});

/**
 * Extracts Variants from an object:
 */
export type TExtractVariants<Styles> = Styles extends {
  variants: infer Variants;
}
  ? { [a in keyof Variants]: keyof Variants[a] }
  : {};

/**
 * Extracts Breakpoint keys from a config
 */
export type BreakPointsKeys<Config extends TConfig> = keyof Config['breakpoints'];

/**
 * Takes a value and if it's one of the string type representations of a boolean ('true' | 'false')
 * it adds the actual boolean values to it
 */
export type CastStringToBoolean<Val> = Val extends 'true' | 'false' ? boolean | 'true' | 'false' : never;
/**
 * adds the string type to number while also preserving autocomplete for other string values
 */
export type CastNumberToString<Val> = Val extends number ? string & {} : never;

/**
 * Takes a variants object and converts it to the correct type information for usage in props
 */
export type VariantASProps<Config extends TConfig, VariantsObj> = {
  [V in keyof VariantsObj]?:
    | CastStringToBoolean<VariantsObj[V]>
    | VariantsObj[V]
    | CastNumberToString<VariantsObj[V]>
    | {
        [B in BreakPointsKeys<Config> | TMainBreakPoint]?:
          | CastStringToBoolean<VariantsObj[V]>
          | VariantsObj[V]
          | CastNumberToString<VariantsObj[V]>;
      };
};

type MergeElementProps<As extends React.ElementType, Props extends object = {}> = Omit<
  React.ComponentPropsWithRef<As>,
  keyof Props
> &
  Props;

/**
 * Types for a styled component which contain:
 * 1. Props of a styled component
 * 2. The compoundVariants function typings
 */
export interface IStyledComponent<ComponentOrTag extends React.ElementType, Variants, Config extends TConfig> {
  /**
   * Props of a styled component
   */
  <As extends React.ElementType = ComponentOrTag>(
    // Merge native props with variant props to prevent props clashing.
    // e.g. some HTML elements have `size` attribute. And when you combine
    // both types (native and variant props) the common props become
    // unusable (in typing-wise)
    props: MergeElementProps<As, VariantASProps<Config, Variants>> & {
      as?: As;
      css?: TCssWithBreakpoints<Config>;
      className?: string;
      children?: any;
    }
  ): any;
  /**
   * Compound Variant typing:
   */
  compoundVariant: (
    compoundVariants: VariantASProps<Config, Variants>,
    possibleValues: TCssWithBreakpoints<Config>
  ) => IStyledComponent<ComponentOrTag, Variants, Config>;
  /**
   * Default props typing:
   */
  defaultProps?: VariantASProps<Config, Variants> & { [k: string]: any };
  /**
   * DisplayName typing:
   */
  displayName?: string;
}

/** Typed css with tokens and breakpoints */
export type TCssWithBreakpoints<Config extends TConfig> = TCssProp<Config> &
  { [key in BreakPointsKeys<Config>]?: TCssProp<Config> };

/** The type for the styles in a styled call */
export type TComponentStylesObject<Config extends TConfig> = TCssWithBreakpoints<Config> & {
  variants?: {
    [k: string]: {
      [s: string]: TCssWithBreakpoints<Config>;
    };
  };
};
/**
 * Types for styled.button, styled.div, etc..
 */
export type TProxyStyledElements<Config extends TConfig> = {
  [key in keyof JSX.IntrinsicElements]: <BaseAndVariantStyles extends TComponentStylesObject<Config>>(
    a: BaseAndVariantStyles | TComponentStylesObject<Config>
  ) => IStyledComponent<key, TExtractVariants<BaseAndVariantStyles>, Config>;
};
/**
 * Styled Components creator Type.
 * ie: styled.div(styles) | styled('div', {styles})
 */
export type TStyled<Config extends TConfig> = {
  // tslint:disable-next-line: callable-types
  <
    TagOrComponent extends keyof JSX.IntrinsicElements | React.ComponentType<any> | IStyledComponent<any, any, Config>,
    BaseAndVariantStyles extends TComponentStylesObject<Config>,
    Variants = TExtractVariants<BaseAndVariantStyles>
  >(
    tag: TagOrComponent,
    baseStyles: BaseAndVariantStyles | TComponentStylesObject<Config>
  ): TagOrComponent extends IStyledComponent<infer T, infer V, Config>
    ? IStyledComponent<T, Omit<V, keyof Variants> & Variants, Config>
    : IStyledComponent<TagOrComponent, Variants, Config>;
} & TProxyStyledElements<Config>;

const createCompoundVariantsMatcher = (breakPoints: any, existingMap?: any) => {
  const map = new Map();
  map.set(MAIN_BREAKPOINT_ID, [...(existingMap?.get(MAIN_BREAKPOINT_ID) || [])]);
  Object.keys(breakPoints).forEach((breakpoint) => map.set(breakpoint, [...(existingMap?.get(breakpoint) || [])]));
  return map;
};

export const createStyled = <Config extends TConfig>(
  config: Config
): {
  css: TCss<Config>;
  styled: TStyled<Config>;
} => {
  const css = createCss(config);
  const defaultElement = 'div';
  const Box = React.forwardRef((props: any, ref: React.Ref<Element>) => {
    const Element = props.as || defaultElement;

    return React.createElement(Element, {
      ref,
      ...props,
      as: undefined,
    });
  });

  let currentAs: string | undefined;

  const configBreakpoints = config.breakpoints || {};

  const styledInstance = (
    baseAndVariantStyles: any = (cssComposer: any) => cssComposer.compose(),
    Component: React.ComponentType<any> = Box
  ) => {
    let numberOfCompoundVariants = 0;
    const as = currentAs;
    const { variants = {}, ...base } = baseAndVariantStyles;
    const baseStyles: any = css(base);
    // compound s vars & constants:
    // keep track of all compound variants:
    const compoundVariants: any[] = [];
    // a map that keeps track of the required number of matching s left for each break point:
    const requiredMatches = createCompoundVariantsMatcher(configBreakpoints);
    // keep track of the number of available variants
    const evaluatedVariantMap: Map<string, Map<string, { [key: string]: string }>> = new Map();
    // store pre evaluated variants
    const evaluatedCompoundVariants: Map<any, { [key: string]: string }> = new Map();

    // tslint:disable-next-line: forin
    for (const Name in variants) {
      const variantMap: Map<string, { [key: string]: string }> = new Map();
      // tslint:disable-next-line: forin
      for (const ValueName in variants[Name]) {
        const evaluatedStyles = evaluateStylesForAllBreakpoints(variants[Name][ValueName], configBreakpoints, css);
        variantMap.set(ValueName, evaluatedStyles);
      }
      evaluatedVariantMap.set(Name, variantMap);
    }

    const stitchesComponentId = `scid-${hashString(JSON.stringify(baseAndVariantStyles))}`;

    const StitchesComponent = React.forwardRef((props: any, ref: React.Ref<Element>) => {
      // Check the memoCompsition's identity to warn the user
      // remove in production
      if (process.env.NODE_ENV === 'development') {
        // we're breaking the rules of hooks on purpose as the env will never change
        // eslint-disable-next-line
        const memoStyled = React.useMemo(() => props.css, []); // We want this to only eval once
        if (memoStyled !== props.css && !hasWarnedInlineStyle) {
          // tslint:disable-next-line
          console.warn(
            '@stitches/react : The css prop should ideally not be dynamic. Define it outside your component using the css composer, or use a memo hook'
          );
          hasWarnedInlineStyle = true;
        }
      }

      const compositions = [baseStyles];

      const propsWithoutVariantsAndCssProp: any = {};
      // clone the compound s matcher
      const compoundRequiredMatches = createCompoundVariantsMatcher(configBreakpoints, requiredMatches);
      // keep track of the number of unResolved s so that we could bail early:
      const numberOfUnResolvedCompoundVariants = {
        current: numberOfCompoundVariants,
      };
      for (const key in props) {
        // check if the prop is a variant
        if (key in variants) {
          const evaluatedVariant = evaluatedVariantMap.get(key);
          // normalize the value so that we only have to deal with one structure:
          const keyVal =
            props[key] && typeof props[key] === 'object' ? props[key] : { [MAIN_BREAKPOINT_ID]: props[key] };
          // tslint:disable-next-line: forin
          for (const breakpoint in keyVal) {
            const stringBreakpointVal = String(keyVal[breakpoint]);
            // check if the variant exist for this breakpoint
            if (evaluatedVariant && evaluatedVariant.get(stringBreakpointVal)) {
              compositions.push(evaluatedVariant.get(stringBreakpointVal)?.[breakpoint]);
            }
            /** Compound variants: */
            if (numberOfUnResolvedCompoundVariants.current) {
              compoundVariants.forEach((compoundVariant, i) => {
                // if this breakpoint  matches a compound
                // eslint-disable-next-line
                if (stringBreakpointVal === String(compoundVariant[key])) {
                  compoundRequiredMatches.get(breakpoint)[i]--;
                }
                // when the required matches reach 0 for any compound ...
                // we know we have a matched compoundVariant
                if (compoundRequiredMatches.get(breakpoint)[i] === 0) {
                  numberOfUnResolvedCompoundVariants.current--;
                  compositions.push(evaluatedCompoundVariants.get(compoundVariant)?.[breakpoint]);
                }
              });
            }
            /** End compound variants */
          }
        } else {
          propsWithoutVariantsAndCssProp[key] = props[key];
        }
      }

      if (propsWithoutVariantsAndCssProp.css) {
        compositions.push(propsWithoutVariantsAndCssProp.css);
        propsWithoutVariantsAndCssProp.css = undefined;
      }

      return React.createElement(Component, {
        ...propsWithoutVariantsAndCssProp,
        as: props.as || as,
        ref,
        className: css(stitchesComponentId, ...compositions, props.className).toString(),
      });
    });
    StitchesComponent.displayName =
      typeof currentAs === 'string'
        ? `styled(${currentAs})`
        : Component && Component.displayName
        ? `styled(${Component.displayName})`
        : `styled(Component\)`;

    StitchesComponent.toString = () => `.${stitchesComponentId}`;

    (StitchesComponent as any).compoundVariant = (compundVariantsObject: any, compoundVariantStyles: any) => {
      // Update component level variables:
      numberOfCompoundVariants++;
      // Each time we add
      compoundVariants.push(compundVariantsObject);
      // required matches is a map with breakpoints
      // each time we add a compound variant, we also push its length to
      // all of the breakpoints so:
      // requiredMatches.get(breakpoint)[i] === Object.keys(compoundVariants[i]).length
      // at render time we clone the requiredMatches map and whenever a prop matches a compound variant we decrement
      // the required matches for this compound variant at this breakpoint
      // when the required matches hit 0 we know it's a mtach
      requiredMatches.forEach((value, key) => {
        value.push(Object.keys(compundVariantsObject).length);
      });

      const evaluatedStyles = evaluateStylesForAllBreakpoints(compoundVariantStyles, configBreakpoints, css);

      evaluatedCompoundVariants.set(compundVariantsObject, evaluatedStyles);
      return StitchesComponent;
    };
    return StitchesComponent;
  };

  // tslint:disable-next-line
  const styledProxy = new Proxy(() => {}, {
    get(_, prop) {
      if (prop === 'Box') {
        return Box;
      }
      currentAs = String(prop);
      return styledInstance;
    },
    apply(_, __, [Element, styling]) {
      if (typeof Element === 'string') {
        currentAs = Element;
        return styledInstance(styling);
      }
      currentAs = undefined;
      return styledInstance(styling, Element);
    },
  });

  return {
    styled: styledProxy as any,
    css,
  };
};
function evaluateStylesForAllBreakpoints(styleObject: any, configBreakpoints: any, css: any) {
  const breakpoints: { [key: string]: string } = {
    [MAIN_BREAKPOINT_ID]: css(styleObject),
  };
  if (configBreakpoints) {
    // tslint:disable-next-line
    for (const breakpoint in configBreakpoints) {
      breakpoints[breakpoint] = css({
        [breakpoint]: styleObject,
      });
    }
  }
  return breakpoints;
}
