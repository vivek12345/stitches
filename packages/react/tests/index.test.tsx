import * as React from 'react';
import * as renderer from 'react-test-renderer';
import { createStyled, _ATOM } from '../src';

/**
 * resolves nested rules for media queries into CSSStyleRules
 */
const resolveRules = (rule: CSSRule, resolvedRules: CSSStyleRule[] = []) => {
  if (rule instanceof CSSStyleRule) {
    resolvedRules.push(rule);
  }

  if (rule instanceof CSSMediaRule) {
    for (let i = 0; i < rule.cssRules.length; i++) {
      resolveRules(rule.cssRules[i], resolvedRules);
    }
  }
  return resolvedRules;
};

/**
 * TODO: improve looping
 */
const getMatchingRulesForClasses = (classesAsArray: string[]) => {
  const orderedAppliedStyles: string[] = [];
  for (let i = 0; i < window.document.styleSheets.length; i++) {
    const sheetRules = window.document.styleSheets[i].cssRules as any;
    for (let b = 0; b < sheetRules.length; b++) {
      const rule = sheetRules[b];
      const resolvedRules = resolveRules(rule);
      resolvedRules.forEach((cssRule) => {
        classesAsArray.forEach((atomClass: any) => {
          if (cssRule.selectorText.includes(atomClass)) {
            orderedAppliedStyles.push(cssRule.cssText);
          }
        });
      });
    }
  }
  return orderedAppliedStyles;
};

let { styled, css } = createStyled({});
const resolveVal = (val: any) => {
  const { className, ...props } = val.props;

  // inject styles by calling to string on the atom
  const injectedClasses = className.toString();
  const classesAsArray = injectedClasses.split(' ').map((atomClass: any) => `.${atomClass}`);
  const orderedAppliedStyles = getMatchingRulesForClasses(classesAsArray);

  return {
    orderedAppliedStyles,
    props: { ...props, className: injectedClasses },
    children: val.children,
  };
};

expect.addSnapshotSerializer({
  serialize(val, config, indentation, depth, refs, printer) {
    return JSON.stringify(resolveVal(val), null, 2);
  },

  test(val) {
    return val && val.props && val.props.className && val.children && val.type;
  },
});

beforeEach(() => {
  const newStyled = createStyled({});
  styled = newStyled.styled;
  css = newStyled.css;
});

afterEach(() => {
  (css as any).dispose();
});

describe('styled', () => {
  test('Renders basic component', () => {
    const H1 = styled('h1', {
      div: {
        fontSize: 18,
      },
    });

    const tree = renderer.create(<H1 className="hi">hello world</H1>).toJSON();

    expect(tree).toMatchSnapshot();
  });

  test('renders component with variant', () => {
    const Button = styled('button', {
      backgroundColor: 'gray',
      variants: {
        variant: {
          red: { backgroundColor: 'red' },
          blue: { backgroundColor: 'blue' },
        },
      },
    });

    expect(renderer.create(<Button>hello world</Button>).toJSON()).toMatchSnapshot();
    expect(renderer.create(<Button variant="red">hello world</Button>).toJSON()).toMatchSnapshot();
    expect(renderer.create(<Button variant="blue">hello world</Button>).toJSON()).toMatchSnapshot();
  });

  test('Handles css prop', () => {
    const Button = styled('button', {
      backgroundColor: 'gray',
    });

    expect(renderer.create(<Button css={{ color: 'red' }}>hello world</Button>).toJSON()).toMatchSnapshot();
  });

  test('Handles css prop', () => {
    const Button = styled('button', {
      backgroundColor: 'gray',
    });

    expect(renderer.create(<Button css={{ color: 'red' }}>hello world</Button>).toJSON()).toMatchSnapshot();
  });

  test('Handles compound variant', () => {
    const Button = styled('button', {
      backgroundColor: 'gray',
      variants: {
        variant: {
          red: {
            backgroundColor: 'red',
          },
        },
        size: {
          1: {
            height: 30,
          },
          2: {
            height: 60,
          },
        },
      },
    }).compoundVariant(
      {
        variant: 'red',
        size: 1,
      },
      {
        backgroundColor: 'compoundColor',
      }
    );

    expect(
      renderer
        .create(
          <Button variant="red" size={2}>
            not compound
          </Button>
        )
        .toJSON()
    ).toMatchSnapshot();
    expect(
      renderer
        .create(
          <Button variant="red" size={1}>
            compound
          </Button>
        )
        .toJSON()
    ).toMatchSnapshot();
  });
  test('It has default displayName when a string based element is passed', () => {
    const Button = styled('button', {});
    expect(Button.displayName).toBe('styled(button)');
  });
  test('It defaults to styled(component) when styling a react component that has no displayName', () => {
    const Component = () => <div></div>;
    const Button = styled(Component, {});
    expect(Button.displayName).toBe('styled(Component)');
  });
  test("It defaults to a displayName that uses the internal styled component's displayName if it has one", () => {
    const Component = () => <div></div>;
    Component.displayName = 'MyFancyComponent';
    const Button = styled(Component, {});
    expect(Button.displayName).toBe(`styled(${Component.displayName})`);
  });

  test('Breakpoints work in variants', () => {
    const { styled: _styled, css: _css } = createStyled({
      showFriendlyClassnames: true,
      breakpoints: {
        breakpointOne: (rule) => `@media(min-width:400px){${rule}}`,
        breakpointTwo: (rule) => `@media(min-width:800px){${rule}}`,
      },
    });
    const Button = _styled('button', {
      color: 'red',
      breakpointOne: {
        height: '10px',
      },
      breakpointTwo: {
        height: '20px',
      },
      variants: {
        size: {
          big: {
            breakpointOne: {
              height: '100px',
            },
          },
        },
      },
    });
    expect(renderer.create(<Button>no variant</Button>).toJSON()).toMatchSnapshot();
    (_css as any).dispose();
  });

  test('Breakpoints work in variants and when the variant is passed to the jsx element', () => {
    const { styled: _styled, css: _css } = createStyled({
      showFriendlyClassnames: true,
      breakpoints: {
        breakpointOne: (rule) => `@media(min-width:400px){${rule}}`,
        breakpointTwo: (rule) => `@media(min-width:800px){${rule}}`,
      },
    });
    const Button = _styled('button', {
      color: 'red',
      breakpointOne: {
        height: '10px',
      },
      breakpointTwo: {
        height: '20px',
      },
      variants: {
        size: {
          small: {
            breakpointOne: {
              height: '100px',
            },
            breakpointTwo: {
              height: '200px',
            },
          },
          big: {
            breakpointOne: {
              height: '1000px',
            },
            breakpointTwo: {
              height: '2000px',
            },
          },
        },
      },
    });
    expect(renderer.create(<Button size="big">with variant</Button>).toJSON()).toMatchSnapshot();
    (_css as any).dispose();
  });

  test('Breakpoints work in variants and when a responsive variant is passed ot the element', () => {
    const { styled: _styled, css: _css } = createStyled({
      showFriendlyClassnames: true,
      breakpoints: {
        breakpointOne: (rule) => `@media(min-width:400px){${rule}}`,
        breakpointTwo: (rule) => `@media(min-width:800px){${rule}}`,
      },
    });
    const Button = _styled('button', {
      color: 'red',
      breakpointOne: {
        height: '10px',
      },
      breakpointTwo: {
        height: '20px',
      },
      variants: {
        size: {
          small: {
            breakpointOne: {
              height: '100px',
            },
            breakpointTwo: {
              height: '200px',
            },
          },
          big: {
            breakpointOne: {
              height: '1000px',
            },
            breakpointTwo: {
              height: '2000px',
            },
          },
        },
      },
    });
    expect(
      renderer.create(<Button size={{ breakpointOne: 'small' }}>with responsive variant</Button>).toJSON()
    ).toMatchSnapshot();
    (_css as any).dispose();
  });
  test('It creates a string className', () => {
    const Component = ({ className }: any) => {
      expect(typeof className).toBe('string');
      return <div>hello</div>;
    };
    const Button = styled(Component, {});
    renderer.create(<Button size={{ breakpointOne: 'small' }}>with responsive variant</Button>);
  });

  test('It handles variants with a numeric value of 0', () => {
    const Button = styled('button', {
      variants: {
        size: {
          0: {
            height: '1px',
          },
        },
      },
    });
    expect(renderer.create(<Button size={0}>height should equal '1px'</Button>).toJSON()).toMatchSnapshot();
  });
});
