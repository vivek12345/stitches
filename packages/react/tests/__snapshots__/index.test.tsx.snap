// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`styled Breakpoints work in variants 1`] = `
{
  "orderedAppliedStyles": [
    "._initial_c_vfarC {color: red;}",
    "._breakpointOne_h_KwWJf {height: 10px;}",
    "._breakpointTwo_h_hiyRrM {height: 20px;}"
  ],
  "props": {
    "className": "_initial_c_vfarC _breakpointOne_h_KwWJf _breakpointTwo_h_hiyRrM scid-kvQUME"
  },
  "children": [
    "no variant"
  ]
}
`;

exports[`styled Breakpoints work in variants and when a responsive variant is passed ot the element 1`] = `
{
  "orderedAppliedStyles": [
    "._initial_c_vfarC {color: red;}",
    "._breakpointOne_h_dDcWdv {height: 100px;}",
    "._breakpointTwo_h_dPiPAc {height: 200px;}"
  ],
  "props": {
    "className": "_breakpointOne_h_dDcWdv _breakpointTwo_h_dPiPAc _initial_c_vfarC scid-cXpAaU"
  },
  "children": [
    "with responsive variant"
  ]
}
`;

exports[`styled Breakpoints work in variants and when the variant is passed to the jsx element 1`] = `
{
  "orderedAppliedStyles": [
    "._initial_c_vfarC {color: red;}",
    "._breakpointOne_h_kgNeVv {height: 1000px;}",
    "._breakpointTwo_h_fptqfg {height: 2000px;}"
  ],
  "props": {
    "className": "_breakpointOne_h_kgNeVv _breakpointTwo_h_fptqfg _initial_c_vfarC scid-cXpAaU"
  },
  "children": [
    "with variant"
  ]
}
`;

exports[`styled Handles compound variant 1`] = `
{
  "orderedAppliedStyles": [
    "._gwFOTp {background-color: red;}",
    "._cDgDLB {height: 60px;}"
  ],
  "props": {
    "className": "_cDgDLB _gwFOTp scid-URbJg"
  },
  "children": [
    "not compound"
  ]
}
`;

exports[`styled Handles compound variant 2`] = `
{
  "orderedAppliedStyles": [
    "._izHeYc {height: 30px;}",
    "._eBRzFG {background-color: compoundColor;}"
  ],
  "props": {
    "className": "_eBRzFG _izHeYc scid-URbJg"
  },
  "children": [
    "compound"
  ]
}
`;

exports[`styled Handles css prop 1`] = `
{
  "orderedAppliedStyles": [
    "._fNPZtH {background-color: gray;}",
    "._vfarC {color: red;}"
  ],
  "props": {
    "className": "_vfarC _fNPZtH scid-dmZJpl"
  },
  "children": [
    "hello world"
  ]
}
`;

exports[`styled Handles css prop 2`] = `
{
  "orderedAppliedStyles": [
    "._fNPZtH {background-color: gray;}",
    "._vfarC {color: red;}"
  ],
  "props": {
    "className": "_vfarC _fNPZtH scid-dmZJpl"
  },
  "children": [
    "hello world"
  ]
}
`;

exports[`styled It handles variants with a numeric value of 0 1`] = `
{
  "orderedAppliedStyles": [
    "._NkcBC {height: 1px;}"
  ],
  "props": {
    "className": "_NkcBC scid-fUwIQk"
  },
  "children": [
    "height should equal '1px'"
  ]
}
`;

exports[`styled Renders basic component 1`] = `
{
  "orderedAppliedStyles": [
    "._dawVaV div {font-size: 18px;}"
  ],
  "props": {
    "className": "hi _dawVaV scid-dIyoRL"
  },
  "children": [
    "hello world"
  ]
}
`;

exports[`styled renders component with variant 1`] = `
{
  "orderedAppliedStyles": [
    "._fNPZtH {background-color: gray;}"
  ],
  "props": {
    "className": "_fNPZtH scid-gFxMiu"
  },
  "children": [
    "hello world"
  ]
}
`;

exports[`styled renders component with variant 2`] = `
{
  "orderedAppliedStyles": [
    "._gwFOTp {background-color: red;}"
  ],
  "props": {
    "className": "_gwFOTp scid-gFxMiu"
  },
  "children": [
    "hello world"
  ]
}
`;

exports[`styled renders component with variant 3`] = `
{
  "orderedAppliedStyles": [
    "._htDnsg {background-color: blue;}"
  ],
  "props": {
    "className": "_htDnsg scid-gFxMiu"
  },
  "children": [
    "hello world"
  ]
}
`;
